cpp/src/stream_compaction/distinct.cu

```diff
@@ -37,18 +32,196 @@
 #include <rmm/exec_policy.hpp>
 
 #include <thrust/copy.h>
-#include <thrust/execution_policy.h>
-#include <thrust/functional.h>
+#include <thrust/distance.h>
+#include <thrust/for_each.h>
 #include <thrust/iterator/counting_iterator.h>
 #include <thrust/iterator/discard_iterator.h>
+#include <thrust/uninitialized_fill.h>
 
 #include <utility>
 #include <vector>
 
 namespace cudf {
 namespace detail {
+
+namespace {
+/**
+ * @brief A functor to perform reduction on row indices.
+ *
+ * A reduction operator will be performed on each group of rows that are compared equal.
+ */
+template <typename MapDeviceView, typename KeyHasher, typename KeyComparator>
+struct reduce_op_gen {
+  MapDeviceView const d_map;
+  KeyHasher const d_hash;
+  KeyComparator const d_eqcomp;
+  size_type* const d_output;
+
+  reduce_op_gen(MapDeviceView const& d_map,
+                KeyHasher const& d_hash,
+                KeyComparator const& d_eqcomp,
+                size_type* const d_output)
+    : d_map{d_map}, d_hash{d_hash}, d_eqcomp{d_eqcomp}, d_output{d_output}
+  {
+  }
+
+  template <duplicate_keep_option keep>
+  auto reduce_op() const
+  {
+    return reduce_index_fn<keep>{*this};
+  }
+
+  /**
+   * @brief The functor used on device for row index reduction.
+   *
+   * This inner functor has only one template argument. That reduces the amount of template
+   * parameters required upon constructing this functor, which happens multiple times. Other
+   * template arguments belong to its parent, which needs to be constructed just once.
+   */
+  template <duplicate_keep_option keep>
```

Comment:
Did you benchmark to see if it was important for `duplicate_keep_option` to be a non-type template parameter instead of a runtime value?